research/flamby/fed_isic2019/fedper/fedper_model.py

```diff
@@ -0,0 +1,83 @@
+from typing import Optional
+
+import torch
+import torch.nn as nn
+from efficientnet_pytorch import EfficientNet
+from efficientnet_pytorch.utils import url_map
+from torch.utils import model_zoo
+
+from fl4health.model_bases.fedper_base import FedPerModel
+from research.flamby.utils import shutoff_batch_norm_tracking
+
+
+def from_pretrained(model_name: str, in_channels: int = 3, include_top: bool = False) -> EfficientNet:
+    # There is a bug in the EfficientNet implementation if you want to strip off the top layer of the network, but
+    # still load the pre-trained weights. So we do it ourselves here.
+    model = EfficientNet.from_name(model_name, include_top=include_top)
+    state_dict = model_zoo.load_url(url_map[model_name])
+    state_dict.pop("_fc.weight")
+    state_dict.pop("_fc.bias")
+    model.load_state_dict(state_dict, strict=False)
+    model._change_in_channels(in_channels)
+    return model
+
+
+class HeadClassifier(nn.Module):
+    """FedPer head module"""
+
+    def __init__(self, stack_output_dimension: int):
+        super().__init__()
+        self.fc1 = nn.Linear(stack_output_dimension, 8)
+        self.dropout = nn.Dropout(0.2)
+
+    def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+        x = self.dropout(input_tensor)
+        x = self.fc1(x)
+        return x
+
+
+class BaseEfficientNet(nn.Module):
+    """FedPer feature extractor module
+    We use the EfficientNets architecture that many participants in the ISIC
+    competition have identified to work best.
+    See here the [reference paper](https://arxiv.org/abs/1905.11946)
+    Thank you to [Luke Melas-Kyriazi](https://github.com/lukemelas) for his
+    [pytorch reimplementation of EfficientNets]
+    (https://github.com/lukemelas/EfficientNet-PyTorch).
+    When loading the EfficientNet-B0 model, we strip off the FC layer to use the model as a feature extractor.
+    We freeze a subset of the layers in order to make sure that FENDA is not training twice as many parameters as the
```

Comment:
Based on the implementation for FedPer, I don't think we're freezing any blocks. To avoid confusion, maybe we change this comment to something like
```
There is an option to freeze a subset of the layers to reduce the number of trainable parameters. However, it is not used in the FedPer experiments.
```